Allow decoupling of display updates from the display link refresh rate for testing

mwyrzykowski opened the pull request:
#### 2143d60f9928cc9f4152f26dfb3f5d13318a0892
<pre>
Allow decoupling of display updates from the display link refresh rate for testing
<a href="https://bugs.webkit.org/show_bug.cgi?id=243372">https://bugs.webkit.org/show_bug.cgi?id=243372</a>

Reviewed by Tim Horton.

Allow decoupling of display updates from the display link refresh rate for testing.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(-[WKOneShotDisplayLinkHandler initWithDrawingAreaProxy:]):
Skip CADisplayLink initialization if using override.

(-[WKOneShotDisplayLinkHandler timerFired]):
Call didRefreshDisplay if the NSTimer fires.

(-[WKOneShotDisplayLinkHandler invalidate]):
Invalidate the timer.

(-[WKOneShotDisplayLinkHandler schedule]):
Create the timer if needed.

(-[WKOneShotDisplayLinkHandler pause]):
Cancel the timer if needed.

Canonical link: <a href="https://commits.webkit.org/253091@main">https://commits.webkit.org/253091@main</a>
</pre>

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -65,10 +68,19 @@ - (id)initWithDrawingAreaProxy:(WebKit::RemoteLayerTreeDrawingAreaProxy*)drawing
         _displayLink.preferredFramesPerSecond = 60;
+#if ENABLE(CUSTOM_DISPLAY_UPDATE_RATES)
+        if (framerateOverride)
+            _updateTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 / framerateOverride target:self selector:@selector(timerFired) userInfo:nil repeats:YES];
```
Looks like both _displayLink and _updateTimer aren't retained. Are we just relying on the fact that they keep a ref count while active?

mwyrzykowski replied:
I'm not sure about the display link but looks like _updateTimer is retained from the recurring nature of the timer. Probably would be more clear if I retained it.

mwyrzykowski replied:
I added retain / release to the updateTimer @smfr but left the displayLink as-is since we don't retain it in the `!ENABLE(CUSTOM_DISPLAY_UPDATE_RATES)` path

hortont424 replied:
The runloop retains the display link while it's installed. Is a bit sketch that we don't keep a reference too, but it's fine.

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -48,6 +48,9 @@
     WebKit::RemoteLayerTreeDrawingAreaProxy* _drawingAreaProxy;
     CADisplayLink *_displayLink;
+#if ENABLE(CUSTOM_DISPLAY_UPDATE_RATES)
+    NSTimer *_updateTimer;
```
This should be a RetainPtr<NSTimer>

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -48,6 +48,9 @@
 @interface WKOneShotDisplayLinkHandler : NSObject {
     WebKit::RemoteLayerTreeDrawingAreaProxy* _drawingAreaProxy;
     CADisplayLink *_displayLink;
+#if ENABLE(CUSTOM_DISPLAY_UPDATE_RATES)
```
How about "TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING" or something

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -65,10 +68,19 @@ - (id)initWithDrawingAreaProxy:(WebKit::RemoteLayerTreeDrawingAreaProxy*)drawing
         _displayLink.preferredFramesPerSecond = 60;
+#if ENABLE(CUSTOM_DISPLAY_UPDATE_RATES)
+        if (framerateOverride)
+            _updateTimer = [[NSTimer scheduledTimerWithTimeInterval:1.0 / framerateOverride target:self selector:@selector(timerFired) userInfo:nil repeats:YES] retain];
```
This should assign to the RetainPtr without a -retain

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -65,10 +69,21 @@ - (id)initWithDrawingAreaProxy:(WebKit::RemoteLayerTreeDrawingAreaProxy*)drawing
         // Note that CADisplayLink retains its target (self), so a call to -invalidate is needed on teardown.
+#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+        NSInteger framerateOverride = [NSUserDefaults.standardUserDefaults integerForKey:@"MainScreenRefreshRate"];
+        _displayLink = framerateOverride ? nil : [CADisplayLink displayLinkWithTarget:self selector:@selector(displayLinkFired:)];
```
A general rule of thumb for clear code is to have the important statements appear only once. So don't repeat the ` [CADisplayLink displayLinkWithTarget:self selector:@selector(displayLinkFired:)];` inside the #ifdef; factor the code so there's just one call.

mwyrzykowski replied:
Sounds good I’ll do that instead

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -65,10 +69,21 @@ - (id)initWithDrawingAreaProxy:(WebKit::RemoteLayerTreeDrawingAreaProxy*)drawing
+#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+        if (framerateOverride) {
+            _updateTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 / framerateOverride target:self selector:@selector(timerFired) userInfo:nil repeats:YES];
+            _updateTimerPaused = YES;
```
Why YES when you've just created a scheduled timer?

mwyrzykowski replied:
That matches the behavior of the `_displayLink`

mwyrzykowski replied:
I could defer creating the timer until it’s needed in `schedule`, initially I tried to mirror the same behavior as CADisplayLink

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -103,20 +118,42 @@ - (void)displayLinkFired:(CADisplayLink *)sender
 {
     _displayLink.paused = NO;
+#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+    _updateTimerPaused = NO;
```
Rather than this, why not just destroy the timer in `pause`, and recreate it here?

mwyrzykowski replied:
I wasn’t sure if there was a 1:1 mapping of `pause` to `schedule` calls but I could invalidate the timer which should remove it from the run loop in `pause` and then re-create the timer here if `![_updateTimer isValid]`

mwyrzykowski replied:
That would require either caching the update rate or reading the NSUserDefault again, the latter should be fine as I don’t think the call is frequent

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -65,6 +68,9 @@ - (id)initWithDrawingAreaProxy:(WebKit::RemoteLayerTreeDrawingAreaProxy*)drawing
         _drawingAreaProxy = drawingAreaProxy;
         // Note that CADisplayLink retains its target (self), so a call to -invalidate is needed on teardown.
+#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+        if (![NSUserDefaults.standardUserDefaults integerForKey:@"MainScreenRefreshRate"])
```
Don't hide flow of control inside #idefs. Give the class a `std::optional<FramesPerSecond> m_overrideFrameRate` data member. Then here have:

```
#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
  NSNumber* overrideRefreshRate = [NSUserDefaults.standardUserDefaults integerForKey:@"MainScreenRefreshRate"];
  NSInteger overrideRefreshRateValue = [overrideRefreshRate integerValue];
  if (overrideRefreshRateValue)
    m_overrideFrameRate = overrideRefreshRateValue;
```

Then you can test `m_overrideFrameRate` in places where you decide to make the display link or timer.

mwyrzykowski commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -65,10 +69,20 @@ - (id)initWithDrawingAreaProxy:(WebKit::RemoteLayerTreeDrawingAreaProxy*)drawing
+        bool createDisplayLink = true;
+#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+        NSInteger overrideRefreshRateValue = [NSUserDefaults.standardUserDefaults integerForKey:@"MainScreenRefreshRate"];
+        if (overrideRefreshRateValue) {
```
This could be `if (NSInteger overrideRefreshRateValue = [NSUserDefaults.standardUserDefaults integerForKey:@"MainScreenRefreshRate"])` but I think its a little obfuscating

hortont424 commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -48,6 +48,10 @@
     CADisplayLink *_displayLink;
+#if ENABLE(TIMER_DRIVEN_DISPLAY_REFRESH_FOR_TESTING)
+    RetainPtr<NSTimer> _updateTimer;
+    std::optional<WebCore::FramesPerSecond> m_overrideFrameRate;
```
No `m` prefix on objc ivars (see all the adjacent ones).

mwyrzykowski committed:
Allow decoupling of display updates from the display link refresh rate for testing
https://bugs.webkit.org/show_bug.cgi?id=243372

Reviewed by Tim Horton.

Allow decoupling of display updates from the display link refresh rate for testing.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(-[WKOneShotDisplayLinkHandler initWithDrawingAreaProxy:]):
Skip CADisplayLink initialization if using override.

(-[WKOneShotDisplayLinkHandler timerFired]):
Call didRefreshDisplay if the NSTimer fires.

(-[WKOneShotDisplayLinkHandler invalidate]):
Invalidate the timer.

(-[WKOneShotDisplayLinkHandler schedule]):
Create the timer if needed.

(-[WKOneShotDisplayLinkHandler pause]):
Cancel the timer if needed.

Canonical link: https://commits.webkit.org/253091@main
